train.py

```diff
@@ -0,0 +1,253 @@
+import uuid
+import os.path
+import itertools
+from typing import Callable
+
+import hydra
+import torch.optim
+import torch.nn as nn
+from tqdm import tqdm
+from torch.utils.data import DataLoader
+from omegaconf import OmegaConf, DictConfig
+
+import wandb
+from model import PitchSeqNN
+from data.dataset import BagOfPitches
+from utils import test_model, make_confusion_matrix
+
+
+@hydra.main(version_base=None, config_path="conf", config_name="config")
+def main(cfg: DictConfig):  # base_acc
+    composers = [
+        ["Alexander Scriabin", "Johann Sebastian Bach"],  # 0.58
+        ["Franz Liszt", "Johann Sebastian Bach"],  # 0.76
+        ["Alexander Scriabin", "Sergei Rachmaninoff"],  # 0.63
+        ["Robert Schumann", "Sergei Rachmaninoff"],  # 0.54
+        ["Robert Schumann", "Ludwig van Beethoven"],  # 0.59
+        ["Alexander Scriabin", "Johann Sebastian Bach", "Franz Liszt", "Ludwig van Beethoven"],
+    ]
+    for classnames in composers:
+        print(classnames)
+        print(test_and_confusion_matrix_main(cfg, classnames))
+
+
+def test_and_confusion_matrix_main(cfg: DictConfig, classnames: list[str]):
+    """
+    Evaluate a neural network model on test data, generate a confusion matrix,
+    and return the validation loss and accuracy.
+
+    Args:
+        cfg (DictConfig): Configuration settings for the experiment.
+        classnames (list[str]): List of class names or labels.
+
+    Returns:
+        Dict: Validation loss and accuracy achieved on the test data.
+    """
+    path = classnames[0].replace(" ", "_").lower()
+    for classname in classnames[1:]:
+        path += f"-{classname.replace(' ', '_').lower()}"
+    path += ".pth"
+    if not os.path.isfile(f"models/{path}"):
+        model = run_experiment(cfg, "MIDI-18-bag-of-pitches-pairs", classnames=classnames)
+        if cfg.model.save_model == "y":
+            torch.save(model.state_dict(), f"models/{path}.pth")
+    else:
+        model = PitchSeqNN(cfg.model.hidden_layers, 128, len(classnames))
+        model.load_state_dict(torch.load(f"models/{path}.pth"))
```

Comment:
I think it would be better to try and get rid of this `if` statement - if the point of this function is "test", there shouldn't be a logical path to train the model inside  of it as well.

If you want to have a single method that takes care of the full experiment for a pair of composers, I think it's better to have more generic name, like "def process_composers()", rather than precisely descriptive, like "def do_A_and_B_and_maybe_C()":

```python
def process_composers(composers: list[str], cfg: DictConfig) -> dict:
  model = train_classifier(composers, cfg)
  validation_results = validate_classifier(model, cfg)
  return validation_results
```

This may require some untangling, but I think it's worth it. In the top function you could do:

```python
scores = []
for composer_pair in composers:
  score = process_composers(composer_pair)
  score["composer_pair"] = composer_pair
  scores.append(score)
```

And use the `scores` object if you want to do any result visualization or processing. But I think this should be used more for a local review and to design better metrics to push to wandb :)